File: Tree-shaped Deep Forest/Binary Search Tree/Q3. Delete Node in a BST/Q3_DeleteNodeinaBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        // Base case: if root is null, return null
        if (root == null) return null;

        // If the key to be deleted is smaller than the root's key,
        // then it lies in the left subtree
        if (key < root.val) {
            root.left = deleteNode(root.left, key);
        }
        // If the key to be deleted is greater than the root's key,
        // then it lies in the right subtree
        else if (key > root.val) {
            root.right = deleteNode(root.right, key);
        }
        // Key is same as root's key, this is the node to be deleted
        else {
            // Node with only one child or no child
            if (root.left == null) return root.right;  // If left is null, return right subtree
            else if (root.right == null) return root.left;  // If right is null, return left subtree

            // Node with two children:
            // Get the inorder successor (smallest in the right subtree)
            TreeNode minNode = findMin(root.right);

            // Copy the inorder successor's value to this node
            root.val = minNode.val;

            // Delete the inorder successor
            root.right = deleteNode(root.right, minNode.val);
        }

        // Return the (possibly updated) root
        return root;
    }

    // Helper function to find the node with the minimum value
    private TreeNode findMin(TreeNode node) {
        // Loop to find the leftmost leaf
        while (node.left != null) {
            node = node.left;
        }
        return node;
    }
}

File: Tree-shaped Deep Forest/Binary Search Tree/Q3. Delete Node in a BST/Q3_DeleteNodeinaBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val          # value of the current node
#         self.left = left        # pointer to the left child
#         self.right = right      # pointer to the right child

class Solution(object):
    def deleteNode(self, root, key):
        """
        :type root: Optional[TreeNode]
        :type key: int
        :rtype: Optional[TreeNode]
        """

        if not root:
            return None  # base case: if root is null, return None

        if key < root.val:
            root.left = self.deleteNode(root.left, key)  # key is in left subtree
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)  # key is in right subtree
        else:
            # node with only one child or no child
            if not root.left:
                return root.right  # return right subtree if left is None
            elif not root.right:
                return root.left   # return left subtree if right is None

            # node with two children: find the inorder successor (smallest in right subtree)
            min_larger_node = self.getMin(root.right)  # get successor
            root.val = min_larger_node.val  # copy successor's value to current node
            root.right = self.deleteNode(root.right, min_larger_node.val)  # delete the successor

        return root  # return updated root

    def getMin(self, node):
        """
        Helper function to find the minimum value node in a subtree.
        """
        while node.left:
            node = node.left  # go to the leftmost node
        return node  # return node with minimum value
